src/seedu/addressbook/storage/AddressBookDecoder.java

```diff
@@ -0,0 +1,123 @@
+package seedu.addressbook.storage;
+
+import static seedu.addressbook.parser.Parser.PERSON_DATA_ARGS_FORMAT;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+
+import seedu.addressbook.data.AddressBook;
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.person.Address;
+import seedu.addressbook.data.person.Email;
+import seedu.addressbook.data.person.Name;
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.person.Phone;
+import seedu.addressbook.data.person.UniquePersonList;
+import seedu.addressbook.data.tag.Tag;
+import seedu.addressbook.storage.StorageFile.StorageOperationException;
+
+/**
+ * Decodes the storage data file into an AddressBook object.
+ */
+public class AddressBookDecoder {
+
+    /**
+     * Decodes the AddressBook in the storage data file.
+     *
+     * @param dataFilePath file which contains the AddressBook to decode.
+     * @return decoded AddressBook.
+     * @throws IllegalValueException if any field is invalid.
+     * @throws StorageOperationException if cannot decode any.
+     * @throws IOException if there is an error reading the file.
+     */
+    public static AddressBook decodeAddressBook(Path dataFilePath)
+            throws IOException, IllegalValueException, StorageOperationException {
+        final List<Person> successfullyDecodedPersons = decodePersonsFromStrings(getLinesInFile(dataFilePath));
+        return new AddressBook(new UniquePersonList(successfullyDecodedPersons));
+    }
+
+    /**
+     * Gets all lines from the file as a list of strings. Line separators are removed.
+     *
+     * @throws IOException if there is an error reading the file.
+     */
+    private static List<String> getLinesInFile(Path dataFilePath) throws IOException {
+        return new ArrayList<>(Files.readAllLines(dataFilePath));
```

Comment:
forgot to mention this but same as the reason before, reading the file shouldn't be the responsibility of decoder as well.